oneoffs/sharp_positions.py

```diff
@@ -0,0 +1,420 @@
+    for position, data in tqdm(positions.items(), desc="featurize"):
+        assert len(data) == num_models, len(data)
+
+        top_moves = Counter([d[1][0] for d in data])
+        top_n_moves = [m for m, c in top_moves.most_common(one_hot_moves)]
+        if len(top_n_moves) < one_hot_moves:
+            top_n_moves.extend([-1] * (one_hot_moves - len(top_n_moves)))
+        assert len(top_n_moves) == one_hot_moves, "pad with dummy moves"
+        pos_top_moves.append(top_n_moves)
+
+        # Eventaully we want
+        # [model 1 position 1 features, m1 p2 features, m1 p3 features, ... ]
+        # [model 2 position 1 features, m2 p2 features, m2 p3 features, ... ]
+        # [model 3 position 1 features, m3 p2 features, m3 p3 features, ... ]
+        # ...
+        # [model m position 1 features, mm p2 features, mm p3 features, ... ]
+
+        # We'll do position selection by joining [model x position_feature]
+
+        feature_columns = []
+        for model, (v, m, mv) in enumerate(data):
+            # Featurization (for each positions):
+            #   * Value (-1 to 1), Bucketed value
+            #   * Cluster all model by top_n moves (X,Y,Z or other)?
+            #     * value of that move for model
+            #   * policy value of top move
+            model_features = []
+
+            model_features.append(2 * v - 1)
+            # NOTE(sethtroisi): Consider bucketize value by value percentiles.
+            value_bucket = np.searchsorted((0.2, 0.4, 0.6, 0.8), v)
+            model_features.extend(one_hot(5, value_bucket))
+
+            # Policy weight for most common X moves (among all models).
+            policy_weights = [0] * (one_hot_moves + 1)
+            for move, policy_value in zip(m, mv):
+                if move in top_n_moves:
+                    policy_weights[top_n_moves.index(move)] = policy_value
+                else:
+                    policy_weights[-1] += policy_value
+            model_features.extend(policy_weights)
+
+            assert len(model_features) == num_features
+
+            feature_columns.append(model_features)
+        features.append(feature_columns)
+
+    features = np.array(features)
+    print("Feature shape", features.shape)
+    print()
+
+    # Split the models to test / train
+    train_size = int(num_models * 0.9)
+    train_models = sorted(np.random.permutation(num_models)[:train_size])
+    test_models = sorted(set(range(num_models)) - set(train_models))
+    assert set(train_models + test_models) == set(range(num_models))
+    features_train = features[:, train_models, :]
+    features_test  = features[:, test_models, :]
+
+    labels_train = model_ratings[train_models]
+    labels_test = model_ratings[test_models]
+
+    # Choose some set of positions and see how well they explain ratings
+    positions_to_use = set(positions.keys())
+    linearSVM = svm.LinearSVR()
+    best_test_r2 = 0
+    below_threshold = 0
+
+    for iteration in itertools.count(1):
```

Comment:
just one iteration?

Reply:
itertools.count() is a infinite loop, the 1 just indicates where to start.